src/VitessceGrid.js

```diff
@@ -41,101 +40,100 @@ function shallowEqual(objA, objB) {
-    // > Do not rely on it to “prevent” a rendering, as this can lead to bugs.
-    // so there is probably a better approach.
-    return !shallowEqual(this.props, nextProps);
-  }
-
-  render() {
-    const {
-      layout, getComponent, padding, margin, draggableHandle,
-      reactGridLayoutProps, onAllReady, rowHeight,
-    } = this.props;
-
-    const ResponsiveGridLayout = WidthProvider(Responsive);
-
-    const {
-      cols, layouts, breakpoints, components,
-    } = resolveLayout(layout);
-
-    // Inline CSS is generally avoided, but this saves the end-user a little work,
-    // and prevents class names from getting out of sync.
-    const style = (
-      <style>
-        {`
+const VitessceGridComponent = ({
+  k, v, Component, onReady, removeGridComponent,
+}) => (
+  <div key={k}>
+    <Component
+      {... v.props}
+      clearComponent={removeGridComponent}
+      onReady={onReady}
+    />
+  </div>
+);
+
+const VitessceGrid = (props) => {
+  const {
+    layout, getComponent, padding, margin, draggableHandle,
+    reactGridLayoutProps, onAllReady, rowHeight,
+  } = props;
+  const ResponsiveGridLayout = WidthProvider(Responsive);
+  const {
+    cols, layouts, breakpoints, components,
+  } = resolveLayout(layout);
+
+  // eslint-disable-next-line no-unused-vars
+  const [readyComponentKeys, setReadyComponentKeys] = useState(new Set());
+  const [gridComponents, setGridComponents] = useState(components);
+
+  // Inline CSS is generally avoided, but this saves the end-user a little work,
+  // and prevents class names from getting out of sync.
+  const style = (
+    <style>
+      {`
           ${draggableHandle} {
             cursor: grab;
           }
           ${draggableHandle}:active {
             cursor: grabbing;
           }
-        `}
-      </style>
-    );
-
-    const layoutChildren = Object.entries(components).map(([k, v]) => {
-      const Component = getComponent(v.component);
-      const onReady = () => {
-        this.setState((state) => {
-          const { readyComponentKeys } = state;
-          const newReadyComponentKeys = new Set(readyComponentKeys);
-          newReadyComponentKeys.add(k);
-          if (newReadyComponentKeys.size === Object.keys(components).length) {
-            // The sets are now equal.
-            onAllReady();
-          }
-          return { readyComponentKeys: newReadyComponentKeys };
-        });
-      };
-      return (
-        <div key={k}>
-          <Component {... v.props} onReady={onReady} />
-        </div>
-      );
+          `}
+    </style>
+  );
+
+  const layoutChildren = Object.entries(gridComponents).map(([k, v]) => {
+    const Component = getComponent(v.component);
+    const onReady = () => {
+      setReadyComponentKeys((prevReadyComponentKeys) => {
+        prevReadyComponentKeys.add(k);
+        if (prevReadyComponentKeys.size === Object.keys(gridComponents).length) {
+          // The sets are now equal
+          onAllReady();
```

Comment:
It feels wrong (is "wrong") to mutate `prevReadyComponentKeys` here, rather than creating a copy. But this carries out the desired behavoir and:

```javascript
    const onReady = () => {
      setReadyComponentKeys((prevReadyComponentKeys) => {
        const nextReadyComponentKeys = new Set(prevReadyComponentKeys);
        nextReadyComponentKeys.add(k);
        if (nextReadyComponentKeys.size === Object.keys(gridComponents).length) {
          // The sets are now equal
          onAllReady();
        }
        return nextReadyComponentKeys;
      });
    };
```
causes an infinite loop...?

Reply:
huh. Good catch. Is the infinite loop in here, or somewhere else? I'm just not seeing what in these few lines would cause infinite recursion, so you probably mean something else?